### PR TITLE
Add TTY-aware color (--color, FORCE_COLOR) and OSC 8 hyperlinks

### LINKED ISSUE #415: TTY-aware color (--color, FORCE_COLOR) and OSC 8 clickable rule ids and paths
skillsaw decides whether to emit ANSI color solely by checking for the `NO_COLOR` environment variable — there is no `isatty()` check on stdout anywhere. `skillsaw lint | less` or redirecting to a file leaks raw `^[[91m` escapes, which is the one place a polished CLI looks broken. The fix is the standard cascade every mature tool ships (`ruff`, `pip`, `gh`): `--color always|never|auto` > `FORCE_COLOR` > `NO_COLOR` > `stream.isatty()`.

Once capability detection exists, the same gate enables OSC 8 hyperlinks: rule ids link to their skillsaw.org docs page and file paths become clickable `file://` links, which lets the redundant "Rule docs" footer collapse in interactive terminals.

## Why

Verified against main (`5d2dbf9`):

- Color is gated only on `NO_COLOR` presence, in exactly two places: `src/skillsaw/formatters/text.py:35` and `_ansi_colors()` at `src/skillsaw/cli/_helpers.py:147`.
- `isatty()` appears only for the stderr progress indicator (`_helpers.py:27`) and stdin prompts in marketplace code — never stdout.
- No `--color` flag in `src/skillsaw/cli/_parser.py`; zero `FORCE_COLOR` or OSC 8 (`\x1b]8;;`) references anywhere in `src/`.
- Live repro: `skillsaw lint | cat -v` shows literal `^[[91m` escapes in the piped output.
- The "Rule docs" footer (`text.py:80`) repeats one docs URL per rule that OSC 8 links would make redundant interactively.

## Proposed implementation

1. Add `color_enabled(stream)` to `src/skillsaw/cli/_helpers.py` implementing the cascade: explicit `--color always|never|auto` (new flag in `src/skillsaw/cli/_parser.py`) > `FORCE_COLOR` > `NO_COLOR` > `stream.isatty()`. Document the chosen precedence in `docs/cli.md`.
2. Thread the resolved bool through `format_output`/`format_text` (`src/skillsaw/formatters/__init__.py`, `src/skillsaw/formatters/text.py`) and the four ANSI-emitting subcommands — `_lint.py`, `_fix.py`, `_explain.py`, `_badge.py` all call `_ansi_colors()` today. Gate stderr notices (`_helpers.py:180`) on `sys.stderr.isatty()` separately from stdout.
3. Emit OSC 8 hyperlinks only when color is enabled **and** `TERM != dumb`: rule ids → their docs page; paths → `file://` URLs. Note: `#L<line>` fragments on `file://` are nonstandard and ignored by many terminals — harmless, but don't advertise line jumping.
4. Collapse the "Rule docs" footer only in hyperlink-capable (TTY) mode; piped output keeps the footer, so anything parsing text output today is unaffected.
5. **Gotcha (main regression risk):** `action.yml` must set `FORCE_COLOR=1`. GitHub Actions stdout is not a TTY, so without it every Actions user silently loses colored logs the release this lands.
6. **Gotcha:** existing tests assert color-by-default through non-TTY capture and must be updated: `tests/test_formatters.py` `test_text_includes_ansi_by_default` (~line 194) and `tests/test_integration.py` `test_custom_rule_warning_colors_respect_no_color` (~line 2562, a subprocess-pipe assert of `\x1b[` on stderr with `NO_COLOR` unset). Convert them to `FORCE_COLOR=1` or invert the expectation.

Files to touch: `src/skillsaw/cli/_parser.py`, `src/skillsaw/cli/_helpers.py`, `src/skillsaw/formatters/text.py`, `src/skillsaw/formatters/__init__.py`, `src/skillsaw/cli/_lint.py`, `src/skillsaw/cli/_fix.py`, `src/skillsaw/cli/_explain.py`, `src/skillsaw/cli/_badge.py`, `action.yml`, `tests/test_formatters.py`, `tests/test_integration.py`, `docs/cli.md`, `README.md`.

## Scope

**Effort: M** — the mechanics are simple, but the bool threads through four subcommands plus formatters, and the action.yml + test updates are mandatory, not optional polish.

Out of scope:

- Color themes or new color choices — same palette, just correct gating.
- ANSI/OSC 8 in JSON, SARIF, HTML, or the new github/markdown formatters — those stay byte-identical.
- Windows legacy console handling (modern Windows terminals enable VT processing by default).

Interactions with open PRs:

- **#409** (fixable markers/fix hints) modifies `formatters/text.py`, `_fix.py`, and `tests/test_formatters.py` — direct overlap; whichever lands second rebases, and the fix-hint styling should route through the same gate.
- **#408** (github/markdown formats) modifies `formatters/__init__.py` dispatch and `action.yml` — both files overlap here; coordinate so the color bool only threads into the text formatter and the `FORCE_COLOR=1` env addition merges cleanly with #408's action changes.
- **#410** no interaction; **#407** is merged (coverage tooling, unrelated). #398/#400/#401 add other flags to `_parser.py` — trivial merge friction only.

---
Surfaced and adversarially verified by a multi-agent codebase exploration. 🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by not-stbenjam**
## What

Makes ANSI color TTY-aware with the standard `--color always|never|auto` > `FORCE_COLOR` > `NO_COLOR` > `isatty()` cascade, and adds OSC 8 terminal hyperlinks (clickable rule ids and file paths) on capable terminals.

Fixes #415

## Why

Color was gated solely on `NO_COLOR` presence — there was no `isatty()` check on stdout anywhere, so `skillsaw lint | less` or redirecting to a file leaked raw `^[[91m` escapes. That is the one place a polished CLI looks broken, and it also meant `--output text:file.txt` wrote ANSI into report files.

## How

- **`color_enabled(stream, mode)`** in `cli/_helpers.py` implements the cascade: explicit `--color` flag > `FORCE_COLOR` (non-empty forces on, `0` forces off, and it outranks `NO_COLOR` so CI setups exporting both get what they asked for) > `NO_COLOR` (present, even empty, disables) > `stream.isatty()`.
- **`--color always|never|auto`** added to the four ANSI-emitting subcommands: `lint`, `fix`, `explain`, `badge`. Each resolves the bool once against `sys.stdout`; `_ansi_colors()` now takes that bool instead of reading the environment.
- **`format_report`/`format_text`** take `color`/`hyperlinks` kwargs and no longer read the environment — `--output text:` file reports are always plain, even with `FORCE_COLOR=1`.
- **stderr notices** (custom-rule warning) gate on `sys.stderr` capability separately from stdout.
- **OSC 8 hyperlinks** (`hyperlinks_enabled`: color AND `isatty()` AND `TERM != dumb`): rule ids link to their skillsaw.org docs page, file paths become clickable `file://` links, and the per-rule "Rule docs" URL footer collapses to a one-line dim hint. The PR #409 fixable markers/fix-hint styling routes through the same gate.
- **`action.yml`** exports `FORCE_COLOR: '1'` for the lint step (one env line, placed to merge cleanly with #408) — Actions stdout is not a TTY, so without it every Actions user would silently lose colored logs.
- **Docs**: cascade + hyperlink section in README and in the generated `docs/cli.md` (via `scripts/generate-site-content.py`).

### Before / after

| Context | Before | After |
|---|---|---|
| TTY | colored, URL footer | colored + clickable rule ids/paths, footer collapsed to one hint line |
| `lint \| less` / `> file` | raw `^[[91m` escapes leak | plain text, byte-compatible with today's `NO_COLOR` output (footer kept) |
| GitHub Actions | colored (NO_COLOR unset) | still colored via `FORCE_COLOR=1` in action.yml; no OSC 8 garbage |
| `--output text:report.txt` | ANSI written into the file | always plain |

Hyperlinks are deliberately never forced through a pipe (even `--color always` / `FORCE_COLOR`): CI log viewers render SGR colors but display raw OSC 8 bytes as garbage, and keeping pipes OSC 8-free keeps piped output stable for anything parsing it. This is a slight tightening of the issue's "color enabled AND TERM != dumb" wording, per its own note that footer collapse is for "hyperlink-capable (TTY) mode".

## Testing

- `make test` — 2524 passed (full suite), including:
  - `tests/test_color.py` (new): cascade precedence units — each level overriding the next, `FORCE_COLOR=0`, empty-value edge cases, `TERM=dumb`, streams without `isatty`
  - `tests/test_formatters.py`: OSC 8 link presence/absence, footer collapse, plain-by-default regression tests (replacing the color-by-default asserts)
  - `tests/test_integration.py::TestColorOutput` (new, `single-plugin` fixture): plain piped output, `FORCE_COLOR=1` through a real subprocess pipe, `FORCE_COLOR` beating `NO_COLOR`, `--color always` into a pipe, `--color never` beating `FORCE_COLOR` and on a pty, real pseudo-tty runs asserting OSC 8 + collapsed footer, `TERM=dumb` suppression, `--output text:` file stays plain
  - stderr notice test converted to the cascade (FORCE_COLOR leg, NO_COLOR leg, piped-plain leg)
- `make lint` clean; `make update` run (docs/cli.md regenerated)
- `skillsaw lint` against openshift-eng/ai-helpers: exit 0, `| cat -v` shows **zero** escape sequences piped; pty run verified by eye (colors, clickable links, collapsed footer)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a shared `--color` option across key CLI commands to control ANSI colors in output.
  * Terminal-friendly hyperlinks now appear in supported terminals for rule references and file locations.

* **Bug Fixes**
  * Output coloring now respects terminal detection and color-related environment settings more consistently.
  * Piped output stays plain, while interactive terminals can show richer formatting.

* **Documentation**
  * Updated CLI and README examples to explain color and hyperlink behavior.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -317,6 +317,17 @@ Violations that `skillsaw fix` can resolve automatically are marked with
...
+### Color and hyperlinks
+
+Terminal output is colored only when stdout is a terminal — piped or
+redirected output stays plain. Override with `--color always|never|auto`,
+which outranks the `FORCE_COLOR` environment variable (non-empty forces
+color on, useful in CI; `0` forces it off), which outranks `NO_COLOR`
+(present disables color). On hyperlink-capable terminals (`TERM` other
+than `dumb`), rule ids link to their documentation pages and file paths
+become clickable `file://` links, replacing the "Rule docs" URL footer
+with a one-line hint.
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply README.md]: Good catch — fixed in 1c5c465. `hyperlinks_enabled()` does gate on `color` (plus a real TTY and `TERM != dumb`), so `--color never`/`NO_COLOR` correctly suppress links. Reworded to "When color is enabled on a real terminal" and noted hyperlinks are never emitted through a pipe.

File: scripts/generate-site-content.py [hunk 2]
```diff
@@ -518,6 +518,29 @@ def generate_rule_page(rule_id, group_name, slug, rules_data, research):
...
+3. `NO_COLOR` — present (even empty) disables color
+4. Otherwise color is used only when stdout is a terminal
+
+Piped or redirected output is plain text by default. On terminals that
+aren't `TERM=dumb`, text output also emits [OSC 8
+hyperlinks](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda):
+rule ids link to their documentation pages, file paths become clickable
+`file://` links, and the per-rule "Rule docs" URL footer collapses to a
+one-line hint. Hyperlinks are never emitted through a pipe, even when
+color is forced.
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply scripts/generate-site-content.py:541]: Fixed in 1c5c465 — tightened `CLI_COLOR_SECTION` in `scripts/generate-site-content.py` to the same "when color is enabled on a real terminal" wording and regenerated `docs/cli.md` so the generated section stays in sync with `hyperlinks_enabled()`.

File: scripts/generate-site-content.py [hunk 3]
```diff
@@ -518,6 +518,30 @@ def generate_rule_page(rule_id, group_name, slug, rules_data, research):
...
 
 
+CLI_COLOR_SECTION = """\
+## Color and hyperlinks
+
+Commands that produce terminal output (`lint`, `fix`, `explain`, `badge`)
+decide whether to emit ANSI colors with the standard cascade, strongest
+first:
+
+1. `--color always` / `--color never`
```
Comment (stbenjam) [inline scripts/generate-site-content.py]: Does tihs match the ergonomics of other parameters? I think not, we have other `--no-` params.

`--no-color` feels more natural, e.g. `--no` for boolean params.

Comment (not-stbenjam) [reply scripts/generate-site-content.py]: Good call — switched to `--[no-]color` in e8c6446 using argparse `BooleanOptionalAction`, so it now matches the `--no-*` family (`--no-baseline`, `--no-custom-rules`, etc.). `--color` forces color on (even through a pipe, like `FORCE_COLOR=1`), `--no-color` forces it off, and no flag = auto-detect. Same tri-state capability as before, just the idiomatic surface. Docs and the color test suite updated.
